File: packages/nodejs/src/domains/oas.ts
import path from 'path';
import {
  InfoObject,
  OpenAPIObject,
  OperationObject,
  PathsObject,
  PathItemObject,
  RequestBodyObject,
  SchemaObject,
} from 'openapi3-ts';
import jsonSchemaRefParser from '@apidevtools/json-schema-ref-parser';
import { match as matchPath } from 'path-to-regexp';
import copy from 'fast-copy';
import deepmerge from 'deepmerge';
import { lint } from '@viron/linter';
import {
  ApiMethod,
  API_METHOD,
  OAS_X_AUTOCOMPLETE,
  OAS_X_PAGES,
  OAS_X_TABLE,
  OAS_X_TAGS,
  OAS_X_THEME,
  OAS_X_THUMBNAIL,
  Theme,
  VironDomains,
  VIRON_DOMAINS,
  XPageContentType,
} from '../constants';
import { getDebug } from '../logging';
import { oasValidationFailure } from '../errors';
import {
  createViewer,
  hasPermission,
  hasPermissionByResourceId,
  isApiMethod,
  method2Permissions,
} from './adminrole';

const debug = getDebug('domains:oas');

export {
  PathsObject as VironPathsObject,
  PathItemObject as VironPathItemObject,
  RequestBodyObject as VironRequestBodyObject,
  SchemaObject as VironSchemaObject,
};

export interface VironOperationObject extends OperationObject {
  operationId: string;
}

export interface VironInfoObjectExtentions {
  [OAS_X_THUMBNAIL]?: string;
  [OAS_X_THEME]?: Theme;
  [OAS_X_TAGS]?: string[];
  [OAS_X_TABLE]?: OasXTable;
  [OAS_X_PAGES]?: OasXPages;
  [OAS_X_AUTOCOMPLETE]?: OasXAutocomplete;
}

export interface VironInfoObject
  extends InfoObject,
    VironInfoObjectExtentions {}

export interface VironOpenAPIObject extends OpenAPIObject {
  info: VironInfoObject;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type OasCustomParameters = Record<string, any>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type OasCustomRequestBody = any;

export interface OasXPageContentAction {
  operationId: string;
  defaultParametersValue?: OasCustomParameters;
  defaultRequestBodyValue?: OasCustomRequestBody;
}

export type OasXPageContentActions = OasXPageContentAction[];

export interface OasXPageContent {
  title: string;
  type: XPageContentType;
  operationId: string;
  defaultParametersValue?: OasCustomParameters;
  defaultRequestBodyValue?: OasCustomRequestBody;
  pagination?: boolean;
  autoRefreshSec?: number;
  actions?: OasXPageContentActions;
  // only used by backend
  resourceId: string;
}

export type OasXPageContents = OasXPageContent[];

export interface OasXPage {
  id: string;
  group: string;
  title: string;
  description: string;
  contents: OasXPageContents;
}

export type OasXPages = OasXPage[];

export interface OasXTablePager {
  requestPageKey: string;
  requestSizeKey: string;
  responseMaxpageKey: string;
  responsePageKey: string;
}

export interface OasXTableSort {
  requestKey: string;
}

export interface OasXTable {
  responseListKey: string;
  pager: OasXTablePager;
  sort: OasXTableSort;
}

export interface OasXAutocomplete {
  responseLabelKey: string;
  responseValueKey: string;
}

export { lint };

// oasを取得
export const get = async (
  oas: VironOpenAPIObject,
  infoExtentions: VironInfoObjectExtentions = {},
  roleIds: string[] = []
): Promise<VironOpenAPIObject> => {
  // viewerが未作成の場合は作成する
  await createViewer(oas);

  // 参照破壊しないようにDeepCopy
  const clonedApiDefinition = copy(oas);
  Object.assign(clonedApiDefinition.info, infoExtentions);

  // x-pages[].contents[]を書き換える
  const rewriteContent = async (
    content: OasXPageContent
  ): Promise<OasXPageContent | null> => {
    const { resourceId, operationId } = content;
    // 権限のないcontentは削除する(nullを返す)
    const { method } = findPathMethodByOperationId(operationId, oas) ?? {};
    if (!method) {
      // contentに書かれているoperationIdが不正
      debug('operation isn`t exists. operationId: %s', operationId);
      return null;
    }
    const tasks = roleIds.map((roleId) =>
      hasPermissionByResourceId(roleId, resourceId, method2Permissions(method))
    );
    for await (const hasPermission of tasks) {
      if (hasPermission) {
        return content;
      }
    }
    return null;
  };

  // x-pages[]を書き換える
  const rewritePage = async (page: OasXPage): Promise<OasXPage | null> => {
    const contents = await Promise.all(page.contents.map(rewriteContent));
    page.contents = contents.filter(Boolean) as OasXPageContents;
    // contentsが0件になった場合はpageごと消す
    return page.contents.length ? page : null;
  };

  // 権限のないOperationをフィルタする
  const filterOperation = async (
    path: string,
    method: ApiMethod
  ): Promise<OperationObject | null> => {
    const pathItem: PathItemObject = clonedApiDefinition.paths[path];
    const operation = pathItem[method];
    if (!operation?.operationId) {
      return operation ?? null;
    }
    const tasks = roleIds.map((roleId) =>
      hasPermission(roleId, path, method, oas)
    );
    for await (const hasPermission of tasks) {
      if (hasPermission) {
        return operation;
      }
    }
    return null;
  };

  // pathItemを書き換える
  const rewritePathItem = async (path: string): Promise<PathsObject | null> => {
    const pathItem: PathItemObject = clonedApiDefinition.paths[path];
    const newPathItems = await Promise.all(
      Object.keys(pathItem).map(async (method: string) => {
        if (!isApiMethod(method)) {
          return null;
        }
        const operation = await filterOperation(path, method);
        if (operation) {
          return { [method]: operation };
        }
        return null;
      })
    );
    const newPathItem = Object.assign({}, ...newPathItems);
    if (Object.keys(newPathItem).length <= 0) {
      return null;
    }
    return { [path]: newPathItem };
  };

  // x-pagesを書き換える
  const pages = await Promise.all(
    (clonedApiDefinition.info[OAS_X_PAGES] ?? []).map(rewritePage)
  );
  clonedApiDefinition.info[OAS_X_PAGES] = pages.filter(Boolean) as OasXPages;

  // pathsを書き換える
  const paths = await Promise.all(
    Object.keys(clonedApiDefinition.paths).map(rewritePathItem)
  );
  clonedApiDefinition.paths = Object.assign({}, ...paths);

  // validation
  const { isValid, errors } = lint(clonedApiDefinition);
  if (!isValid) {
    debug('OAS validation failure. errors:');
    (errors ?? []).forEach((error, i) => debug('%s: %o', i, error));
    debug('oas: %s', JSON.stringify(clonedApiDefinition));
    throw oasValidationFailure();
  }
  return clonedApiDefinition;
};

// oasファイルのパスを取得
export const getPath = (name: VironDomains): string => {
  return path.resolve(__dirname, '..', 'openapi', `${name}.yaml`);
};

// oasをロードして$refを解決する
export const loadResolvedOas = async (
  path: string
): Promise<VironOpenAPIObject> => {
  return (await jsonSchemaRefParser.dereference(path, {
    dereference: {
      circular: false, // 循環参照を許容しない
    },
  })) as VironOpenAPIObject;
};

// oasのパス定義をexpree形式に書き換える
const toExpressStylePath = (path: string): string =>
  path.replace(/{/g, ':').replace(/}/g, '');

// uriにヒットするpathとpathItemを取得する
export const getPathItem = (
  uri: string,
  oas: VironOpenAPIObject
): {
  path: string | null;
  pathItem: PathItemObject | null;
  params: Record<string, string> | null;
} => {
  const { paths } = oas;
  let matchObj: { params: Record<string, string> } | undefined;
  const matchedPath = Object.keys(paths).find((p: string) => {
    const matcher =
      typeof paths[p]['x-viron-path-matcher'] === 'function'
        ? paths[p]['x-viron-path-matcher']
        : matchPath(toExpressStylePath(p));
    paths[p]['x-viron-path-matcher'] = matcher; // cache
    const m = matcher(uri);
    if (m) {
      matchObj = m;
    }
    return !!m;
  });
  return {
    path: matchedPath ?? null,
    pathItem: matchedPath ? paths[matchedPath] : null,
    params: matchObj ? matchObj.params : null,
  };
};

// oasのx-pages内のcontentsを一覧取得
const listContentsByOas = (oas: VironOpenAPIObject): OasXPageContents => {
  const xPages = oas.info[OAS_X_PAGES] ?? [];
  return xPages.map((xPage) => xPage.contents).flat();
};

// x-pagesからoperationIdに一致するresourceIdを取得
const findResourceId = (
  operationId: string,
  oas: VironOpenAPIObject
): string | null => {
  const contents = listContentsByOas(oas);
  const resourceId = contents.find(
    (content) => content.operationId === operationId
  )?.resourceId;
  return resourceId ?? null;
};

// oasの$refを解決する
export const dereference = async (
  oas: VironOpenAPIObject
): Promise<VironOpenAPIObject> => {
  return (await jsonSchemaRefParser.dereference(oas)) as VironOpenAPIObject;
};

// uri,methodに対応するopeartionを取得
export const findOperation = (
  uri: string,
  method: string,
  oas: VironOpenAPIObject
): VironOperationObject | null => {
  const { pathItem } = getPathItem(uri, oas);
  return pathItem?.[method.toLocaleLowerCase()] ?? null;
};

// uri,methodに対応するopeartionIdを取得
const findOperationId = (
  uri: string,
  method: ApiMethod,
  oas: VironOpenAPIObject
): string | null => {
  const { pathItem } = getPathItem(uri, oas);
  const operation = pathItem?.[method.toLocaleLowerCase()] ?? null;
  return operation?.operationId ?? null;
};

interface PathMethod {
  path: string;
  method: ApiMethod;
}

type OperationIdPathMethodMap = Record<string, PathMethod>;

// operationIdからpathとmethodを逆引きするためのマップを生成
let operationIdPathMethodMap: OperationIdPathMethodMap | null;
const genOperationIdPathMethodMap = (
  oas: VironOpenAPIObject
): OperationIdPathMethodMap | null => {
  const { paths } = oas;
  Object.keys(paths).forEach((path) => {
    Object.keys(paths[path]).forEach((method: string) => {
      if (!isApiMethod(method)) {
        return;
      }
      const operationObject = paths[path][method];
      if (operationObject.operationId) {
        operationIdPathMethodMap = operationIdPathMethodMap ?? {};
        operationIdPathMethodMap[operationObject.operationId] = {
          path,
          method,
        };
      }
    });
  });
  return operationIdPathMethodMap;
};

// operationIdからpathとmethodを逆引き
const findPathMethodByOperationId = (
  operationId: string,
  oas: VironOpenAPIObject
): PathMethod | null => {
  const map = operationIdPathMethodMap ?? genOperationIdPathMethodMap(oas);
  return map?.[operationId] ?? null;
};

// operationIdをactionsに持つcontentのresourceIdを取得
const findResourceIdByActions = (
  operationId: string,
  oas: VironOpenAPIObject
): string | null => {
  const contents = listContentsByOas(oas);
  const content = contents.find((content) =>
    (content.actions ?? []).find((action) => action.operationId === operationId)
  );
  return content?.resourceId ?? null;
};

export const clearCache = (): void => {
  operationIdPathMethodMap = null;
};

// uriとmethodからリソースIDを取得する
export const getResourceId = (
  uri: string,
  method: ApiMethod,
  oas: VironOpenAPIObject
): string | null => {
  // operationIdからresourceIdを取得できれば終了
  const operationId = findOperationId(uri, method, oas);
  if (!operationId) {
    return null;
  }

  const resourceId = findResourceId(operationId, oas);
  if (resourceId) {
    debug(
      'Hit the passed uri and method. %s:%s, ResourceId: %s',
      method,
      uri,
      resourceId
    );
    return resourceId;
  }

  /**
   * リクエストURIが `/foo/{id}/bar` のとき
   * `/foo/{id}/bar` > `/foo/{id}` > `/foo` の順にresourceIdを検索する
   */
  let lastIndex, parentResourceId;
  let parentUri = uri;
  do {
    parentUri = parentUri.slice(0, lastIndex);
    Object.values(API_METHOD).some((method) => {
      const oid = findOperationId(parentUri, method, oas);
      parentResourceId = oid ? findResourceId(oid, oas) : null;
      return !!parentResourceId;
    });
    if (parentResourceId) {
      debug(
        'Hit parent uri. %s:%s, ResourceId: %s',
        method,
        uri,
        parentResourceId
      );
      break;
    }
  } while ((lastIndex = parentUri.lastIndexOf('/')) > 0);
  if (parentResourceId) {
    return parentResourceId;
  }

  // uriとmethodがどこかのactionsに定義されているかもしれないので探す
  const actionResourceId = findResourceIdByActions(operationId, oas);
  if (actionResourceId) {
    debug(
      'Hit actions uri. %s:%s, ResourceId: %s',
      method,
      uri,
      actionResourceId
    );
    return actionResourceId;
  }

  // ここまでヒットしなければresourceIdを特定できないのでnullを返す
  return null;
};

// 複数のoasをマージする
export const merge = (
  ...apiDefinitions: VironOpenAPIObject[]
): VironOpenAPIObject => {
  return deepmerge.all(apiDefinitions) as VironOpenAPIObject;
};

// @viron/libが提供している機能のoasを1つにして返す
export const getVironSpec = async (): Promise<VironOpenAPIObject> => {
  const names = Object.values(VIRON_DOMAINS);
  const specs = await Promise.all(
    names.map((name) => loadResolvedOas(getPath(name)))
  );
  return merge(...specs);
};

File: packages/golang/domains/oas.go
package domains

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"

	"github.com/cam-inc/viron/packages/golang/logging"

	"github.com/cam-inc/viron/packages/golang/constant"
	"github.com/cam-inc/viron/packages/golang/helpers"
	"github.com/getkin/kin-openapi/openapi3"
)

type (
	pathMethod struct {
		path   string
		method string
	}

	operationIDPathMethodMap map[string]*pathMethod

	Action struct {
		OperationID string `json:"operationId"`
	}

	Content struct {
		OperationID string    `json:"operationId"`
		Query       []string  `json:"query,omitempty"`
		Sort        []string  `json:"sort,omitempty"`
		ResourceID  string    `json:"resourceId"`
		ContentType string    `json:"type"`
		Actions     []*Action `json:"actions"`
	}

	XPage struct {
		ID          string     `json:"id"`
		Title       string     `json:"title"`
		Group       string     `json:"group"`
		Description string     `json:"description"`
		Contents    []*Content `json:"contents"`
	}

	XPages []*XPage

	Permission struct {
		ResourceID   string
		OperationIDs []string
	}
)

// GetOas ロールに沿ったoasを返す
func GetOas(apiDef *openapi3.T, roleIDs []string) *openapi3.T {
	log := logging.GetDefaultLogger()
	clone := &openapi3.T{}
	if buf, err := json.Marshal(apiDef); err != nil {
		return nil
	} else {
		if err := json.Unmarshal(buf, clone); err != nil {
			return nil
		}
	}

	log.Debugf("roleIDs %+v", roleIDs)

	if err := CreateViewerRole(clone); err != nil {
		log.Errorf("CreateViewerRole error(%+v)", err)
	}

	extentions := helpers.ConvertExtentions(clone)
	rewritedPages := []*helpers.XPage{}
	for _, page := range extentions.XPages {
		granted := []*helpers.XContent{}
		for _, content := range page.Contents {
			pathMethod := findPathMethodByOperationID(content.OperationID, clone)
			if pathMethod == nil || pathMethod.method == "" {
				continue
			}

			log.Debugf("findPathMethodByOperationID %+v", pathMethod)

			for _, roleID := range roleIDs {
				log.Debugf("roleId %s resourceId %s method2Permissions(pathMethod.method) %+v", roleID, content.ResourceID, method2Permissions(pathMethod.method))
				if hasPermissionByResourceID(roleID, content.ResourceID, method2Permissions(pathMethod.method)) {
					granted = append(granted, content)
				}
			}

		}
		log.Debugf("granted %+v", granted)

		if len(granted) > 0 {
			page.Contents = granted
			rewritedPages = append(rewritedPages, page)
		}
	}

	log.Debugf("rewritedPages %+v", rewritedPages)

	clone.Info.Extensions[constant.OAS_X_PAGES] = rewritedPages
	return clone
}

// ACLAllow ロールに沿ったAPIアクセス許可チェック
func ACLAllow(method, uri string, roleIDs []string, apiDef *openapi3.T) bool {

	log := logging.GetDefaultLogger()

	operationID := findOperationIDByPathMethod(method, uri, apiDef)
	if operationID == "" {
		log.Debugf("not operationID uri[%+v] method[%+v] roleIDs[%+v] apiDef[%+v]", uri, method, roleIDs, apiDef)
		return false
	}
	resourceID := findResourceID(operationID, apiDef)
	if resourceID == "" {
		log.Debugf("not resourceID uri[%+v] method[%+v] roleIDs[%+v] apiDef[%+v]", uri, method, roleIDs, apiDef)
		return false
	}
	pathMethod := findPathMethodByOperationID(operationID, apiDef)
	if pathMethod == nil {
		log.Debugf("not pathMethod uri[%+v] method[%+v] roleIDs[%+v] apiDef[%+v]", uri, method, roleIDs, apiDef)
		return false
	}

	for _, roleID := range roleIDs {
		if hasPermissionByResourceID(roleID, resourceID, method2Permissions(pathMethod.method)) {
			return true
		}
	}

	log.Debugf("not permission uri[%+v] method[%+v] roleIDs[%+v] apiDef[%+v]", uri, method, roleIDs, apiDef)
	return false
}

// listContentByOas oasのx-pages内のcontentsを一覧取得
func listContentsByOas(apiDef *openapi3.T) []*Content {
	contents := []*Content{}
	if apiDef.Info == nil || len(apiDef.Info.Extensions) == 0 {
		return contents
	}

	xPages := &XPages{}
	encodedXPages, err := json.Marshal(apiDef.Info.ExtensionProps.Extensions[constant.OAS_X_PAGES])
	if err != nil {
		log.Errorf("x-pages json.marshal failed. err:%v\n", err)
		return contents
	}
	err = json.Unmarshal(encodedXPages, xPages)
	if err != nil {
		log.Errorf("json.Unmarshal failed. err:%v\n", err)
		return contents
	}

	for _, xPage := range *xPages {
		for _, v := range xPage.Contents {
			contents = append(contents, v)
		}
	}
	return contents
}

func genOperationIDPathMethodMap(apiDef *openapi3.T) operationIDPathMethodMap {
	oMap := operationIDPathMethodMap{}
	for path, pathItem := range apiDef.Paths {
		for _, method := range constant.API_METHODS {
			ope := pathItem.GetOperation(helpers.MethodNameUpper(method))
			if ope == nil {
				continue
			}
			oMap[ope.OperationID] = &pathMethod{
				path:   path,
				method: method,
			}
		}
	}
	return oMap
}

// findPathMethodByOperationID operationIdからpathとmethodを逆引き
func findPathMethodByOperationID(operationID string, apiDef *openapi3.T) *pathMethod {
	oMap := genOperationIDPathMethodMap(apiDef)
	pm, ok := oMap[operationID]
	if !ok {
		return nil
	}
	return pm
}

// findOperationIDByPathMethod pathとmethodからoperationIDを取得
func findOperationIDByPathMethod(method, path string, apiDef *openapi3.T) string {
	oMap := genOperationIDPathMethodMap(apiDef)
	for operationID, pathMethod := range oMap {
		// rootは除外
		if pathMethod.path == "/" {
			continue
		}

		// /users/{userId} を /users/+ に正規表現パターンにする
		rep := regexp.MustCompile(`{.+}`)
		pattern := rep.ReplaceAllString(pathMethod.path, "+")
		matcher := regexp.MustCompile(pattern)

		// メソッド名が同じ かつ パスと正規表現でマッチする かつ "/"の数が同じ
		if strings.EqualFold(pathMethod.method, method) && matcher.MatchString(path) && (strings.Count(path, "/") == strings.Count(pathMethod.path, "/")) {
			return operationID
		}
	}
	return ""
}

// genPermissions resourceIDをキーにして、operationIDのlistを返す
func genPermissions(apiDef *openapi3.T) []*Permission {
	var permissions []*Permission
	// x-pagesのcontentsをpermissionの基準にする
	contents := listContentsByOas(apiDef)

	// 全てのcontentの関連するresourceIDとoperationIDのセットをapiDefから収集する
	for _, content := range contents {
		// 基準となるoperationIDに関連するoperationIDの配列を取得
		operationIDs := findPermissionOperationIDs(content.OperationID, apiDef)
		// actionsのoperationIDも同じresourceIDとする
		for _, action := range content.Actions {
			operationIDs = append(operationIDs, action.OperationID)
		}

		permissions = append(permissions, &Permission{
			ResourceID:   content.ResourceID,
			OperationIDs: operationIDs,
		})
	}

	return permissions
}

// findPermissionOperationIDs 指定のoperationIDに関連するoperationIDの配列を返す
func findPermissionOperationIDs(operationID string, apiDef *openapi3.T) []string {
	pathMethod := findPathMethodByOperationID(operationID, apiDef)

	var operationIDs []string
	for path, pathItem := range apiDef.Paths {
		// 対象のpathかどうかをチェック
		// ex)
		// 対象のpath "/users
		// 対象のpath "/users/{xxx}"
		// 非対象のpath "/users/purchases/{xxx}"
		// 非対象のpath "/users/{xxx}/purchases/{xxx}"
		if path == pathMethod.path || (strings.Count(path, "/") == 2 && regexp.MustCompile(fmt.Sprintf(`^%s/{.+}$`, pathMethod.path)).MatchString(path)) {
			operations := pathItem.Operations()
			for _, operation := range operations {
				operationIDs = append(operationIDs, operation.OperationID)
			}
		}
	}
	return operationIDs
}

// PermissionsからresourceIDを取得
func findResourceID(operationID string, apiDef *openapi3.T) string {
	permissions := genPermissions(apiDef)

	for _, permission := range permissions {
		operationIDs := helpers.StringSlice(permission.OperationIDs)
		if operationIDs.Contains(operationID) {
			return permission.ResourceID
		}
	}

	return ""
}
